### ISSUE TITLE
typed-array-buffer-schema vs. protobuf

### ISSUE THREAD

**Issue by bennlich** [geckosio/typed-array-buffer-schema#3]
I'm wondering what the difference is between this lib and google's [protobuf](https://github.com/protocolbuffers/protobuf/tree/master/js). I think the goals are similar, but I'm guessing this is a lighter weight implementation?

**Comment by bennlich** [comment]
I did a little more research and found [PSON](https://github.com/dcodeIO/PSON) and [MessagePack](https://msgpack.org/). There are way more JavaScript implementations of binary wire formats than I thought!

The tradeoffs with each lib are not super clear to me yet, but it seems like the important things to look at are:
- message encoding time
- message decoding time
- message size
- library ease of use

**Comment by yandeu** [comment]
I have never used another library than this.

- What are your experiences?
- Have you done some benchmarks?
- Which package are you using? And why?

**Comment by bennlich** [comment]
Here are my experiences so far:

First I tried typed-array-buffer-schema because it looked simple, and I like geckos. But then I realized that it's pretty strict about the kinds of shapes your data can be in (see https://github.com/geckosio/typed-array-buffer-schema/pull/5), and that was a bummer, so I decided to try out what I thought might be a more mature lib.

I tried PSON next because I liked that it accepts arbitrary JSON shapes, and has a "progressive" mode where, instead of the client and server both knowing the message format ahead of time, the first message it sends describes the schema, and the client learns from that. But this progressive mode didn't turn out to be useful to me (making it useful would require complicating the way I was keeping track of client connections), and the PSON message sizes were much bigger than I wanted (I think because it doesn't support particularly small ints).

I decided that maybe the typed-array-buffer-schema message shape limitations were a good thing, now that I understood them, so I went back to that. I like that it's simple, and because of that, the message sizes are really as small as possible, and the source code is readable. I *think* there's room to improve the allowed shapes without sacrificing message size, but I haven't thought about it too deeply.

I've heard that protobufs are robust, but I haven't tried them out yet. The docs made them seem more complicated to get up and running.

I didn't get around to trying MessagePack yet because typed-array-buffer-schema is meeting my needs so far.
